File: BSvsZP-Common/Common/ComponentInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Runtime.Serialization;

namespace Common
{
    [DataContract]
    public class ComponentInfo : DistributableObject
    {
        
        #region Private Data Members
        // Define this, the Message class, identifier
        private static Int16 ClassId { get { return (Int16)DISTRIBUTABLE_CLASS_IDS.ComponentInfo; } }

        private Int16 id;
        private EndPoint communicationEndPoint;
        #endregion

        #region Public Properties and Other Stuff

        [DataMember]
        public Int16 Id
        {
            get { return id; }
            set
            {
                id = value;
                RaiseChangedEvent();
            }
        }
        [DataMember]
        public EndPoint CommunicationEndPoint
        {
            get { return communicationEndPoint; }
            set
            {
                communicationEndPoint = value;
                RaiseChangedEvent();
            }
        }
        
        public static int MinimumEncodingLength
        {
            get
            {
                return 4                // Object header
                       + 2              // Id
                       + 1;             // CommunicationEndPoint
            }
        }

        public event StateChangeHandler Changed;

        #endregion
      
        #region Constructors
        public ComponentInfo() {}

        public ComponentInfo(Int16 id)
        {
            Id = id;
        }

        public ComponentInfo(Int16 id, EndPoint endPoint)
            : this(id)
        {
            CommunicationEndPoint = endPoint;
        }
        #endregion

        #region Encoding and Decoding methods

        /// <summary>
        /// This method encodes an object of this class into a byte list
        /// </summary>
        /// <param name="bytes"></param>
        public override void Encode(ByteList bytes)
        {
            bytes.Add(ClassId);                             // Write out the class type

            Int16 lengthPos = bytes.CurrentWritePosition;   // Get the current write position, so we
                                                            // can write the length here later

            bytes.Add((Int16) 0);                           // Write out a place holder for the length

            bytes.AddObjects(Id, CommunicationEndPoint);

            Int16 length = Convert.ToInt16(bytes.CurrentWritePosition - lengthPos - 2);
            bytes.WriteInt16To(lengthPos, length);          // Write out the length of this object        
        }

        /// <summary>
        /// This method decodes a message from a byte list.  It can onlt be called from within the class hierarchy.
        /// </summary>
        /// <param name="messageBytes"></param>
        protected override void Decode(ByteList bytes)
        {
            if (bytes == null || bytes.RemainingToRead < MinimumEncodingLength)
                throw new ApplicationException("Invalid byte array");
            else if (bytes.PeekInt16() != ClassId)
                throw new ApplicationException("Invalid class id");
            else
            {
                Int16 objType = bytes.GetInt16();
                Int16 objLength = bytes.GetInt16();

                bytes.SetNewReadLimit(objLength);

                id = bytes.GetInt16();
                communicationEndPoint = bytes.GetDistributableObject() as EndPoint;
                RaiseChangedEvent();

                bytes.RestorePreviosReadLimit();
            }
        }

        #endregion

        #region Event raising methods
        protected void RaiseChangedEvent()
        {
            if (Changed != null)
                Changed(new StateChange() { Type = StateChange.ChangeType.UPDATE, Subject = this });
        }
        #endregion
    }
}

File: BSvsZP-Common -Java/Common/ComponentInfo.java
package Common;

import java.net.UnknownHostException;
import org.omg.CORBA.portable.ApplicationException;
import Common.StateChange.StateChangeHandler;

public class ComponentInfo extends DistributableObject {

    private static short ClassId = (short) DISTRIBUTABLE_CLASS_IDS.ComponentInfo.getValue();
    private short Id;
    private EndPoint CommunicationEndPoint;
    private static int MinimumEncodingLength;
    public StateChangeHandler handler;

    public ComponentInfo() {
    }

    public ComponentInfo(short id) {
        Id = id;
    }

    public ComponentInfo(short id, EndPoint endPoint) {
        this(id);
        setCommmunicationEndPoint(endPoint);
        printAll();
    }

    private void printAll() {
        getClassId();
        getMinimumEncodingLength();

    }

    /*// new
     public static ComponentInfo Create(ByteList bytes) throws ApplicationException, Exception
     {
     ComponentInfo result = new ComponentInfo();
     result.Decode(bytes);
     return result;
     }
     */
    @Override
    public void Encode(ByteList bytes) throws UnknownHostException, Exception {
        bytes.Add((short) DISTRIBUTABLE_CLASS_IDS.ComponentInfo.getValue());                             // Write out the class type
        bytes.update();
        short lengthPos = bytes.getCurrentWritePosition();   // Get the current write position, so we
        // can write the length here later

        bytes.Add((short) 0);                           // Write out a place holder for the length
        bytes.update();
        bytes.AddObjects(getId(), getCommunicationEndPoint()); //bytes.AddObjects( (byte) AgentType.getValue(), getId());
        bytes.update();
        short length = (short) (bytes.getCurrentWritePosition() - lengthPos - 2);
        bytes.WriteInt16To(lengthPos, length);          // Write out the length of this object        
    }

    @Override
    protected void Decode(ByteList bytes) throws ApplicationException, Exception {
        if (bytes == null || bytes.getRemainingToRead() < getMinimumEncodingLength()) {
            throw new ApplicationException("Invalid byte array", null);
        } else if (bytes.PeekInt16() != (short) DISTRIBUTABLE_CLASS_IDS.ComponentInfo.getValue()) {
            throw new ApplicationException("Invalid class id", null);
        } else {
            short objType = bytes.GetInt16();
            short objLength = bytes.GetInt16();
            bytes.update();
            bytes.SetNewReadLimit(objLength);
            bytes.update();

            setId((bytes.GetInt16()));
            bytes.update();
            setCommmunicationEndPoint(((EndPoint) bytes.GetDistributableObject()));
            bytes.update();

            RaiseChangedEvent();

            bytes.RestorePreviosReadLimit();
            bytes.update();
        }
    }

    public short getId() {
        return Id;
    }

    public void setId(short id) {
        Id = id;
        RaiseChangedEvent();
        System.out.println("ComponentInfo.Id " + Id);
    }

    public EndPoint getCommunicationEndPoint() {
        return CommunicationEndPoint;
    }

    public void setCommmunicationEndPoint(EndPoint commmunicationEndPoint) {
        CommunicationEndPoint = commmunicationEndPoint;
        RaiseChangedEvent();
    }

    public static int getMinimumEncodingLength() {

        MinimumEncodingLength = 4 // Object header
                + 2 // Id
                + 1;   					// CommunicationEndPoint
                				/*	+ 1          		  	// Agent Types
         + 1 					//EndPoint.MinimumEncodingLength
         + 1; 					//Tick.MinimumEncodingLength;
         */ System.out.println("ComponentInfo.MinimumEncodingLength =" + MinimumEncodingLength);
        return MinimumEncodingLength;
    }

    public short getClassId() {
        ClassId = (short) DISTRIBUTABLE_CLASS_IDS.ComponentInfo.getValue();
        System.out.println("ComponentInfo.ClassId " + ClassId);
        return ClassId;
    }

    protected void RaiseChangedEvent() {
        if (handler != null) {
            handler.addStateChangeHandler();
        }
    }
}
